### DISCUSSION TITLE
why does "keywords" section in package.json has "socket.io" while soketi actually use "uWebSockets.js" ?

### DISCUSSION THREAD

**Discussion by qiulang** [soketi/pws#817] [General]
I notice this in [package.json](https://github.com/soketi/soketi/blob/master/package.json) 

```
"keywords": [
    "laravel",
    "socket.io",
    "broadcasting",
    "events",
    "redis",
    "socket",
    "pusher"
  ],
...

"dependencies": {
      "uWebSockets.js": "https://github.com/uNetworking/uWebSockets.js.git#v20.10.0",
 }
```

I briefly check the repo history and find that soketi seems to use "uWebSockets.js" from the very beginning, correct me if I am wrong. 

So I have 2 questions here,

1. Why chose "uWebSockets.js" over "socket.io" ?  I search the discussions and only find this https://github.com/soketi/soketi/discussions/314 with the words "Socket.IO is not compatible with this app" 
2. Is "socket.io" in "keywords" a mistake or something else?

Thanks

**Comment by nickchomey** [comment]
I'm not the dev, but I presume that socket.io is included in the keywords so that when people search for socket.io (which is quite popular), soketi will show up in the results.

And uwebsockets was chosen because it is far more performant than sockets.io (which, from what I gather, is actually a TERRIBLE sockets mechanism). You can read more about uwebsockets here 

https://github.com/uNetworking/uWebSockets.js

And the various blog posts here: https://unetworkingab.medium.com/

**Reply by qiulang** [reply]
Why do you think socket.io is a  TERRIBLE sockets mechanism ? I am not challenging you but curious.

I know [uWebSockets.js](https://github.com/uNetworking/uWebSockets.js). His author claimed its performance is much better than socket.io. But other than his author I have yet to see others stick to that claim.

**Reply by nickchomey** [reply]
I've seen this claim made in many places, backed up by benchmarks. Believe what you like - that's the rationale for these choices in soketi.

**Reply by qiulang** [reply]
Can you list one of those places other than the author' s claim? [uWebSockets.js](https://github.com/uNetworking/uWebSockets.js) is implemented in c++, so it is not totally fair to compare them, but no matter what, I would like to who said that, other than the author. Besides, I noticed soketi said that too.

**Reply by nickchomey** [reply]
I don't recall. Google should be helpful.

**Reply by nickchomey** [reply]
I got curious and looked into this again, which just confirmed why I decided to use soketi to begin with 

By all accounts, socket.io is very easy to use, but is slow, bloated and largely unnecessary now that websockets are ubiquitous in browsers. Lower level websockets implementations, like uwebsockets, are far faster and more scalable, but more difficult to implement. 

Yet, some people seem to say that if you're asking the question (you don't really know much about any of this and are building a tool that likely won't ever even reach enterprise scale), you should just use socket.io.

However, that's where soketi comes in. It adds useful layers (eg Pusher API) on top of uwebsockets, can be easily deployed, even on cloudflare workers.

Here's some links I found useful in digging into this again.

* https://www.reddit.com/r/node/comments/l2goiw/is_socketio_still_the_goto_for_websockets_or_is/
* https://ably.com/topic/socketio-vs-websocket
* https://ably.com/topic/socketio
* https://ably.com/topic/websockets
* https://hashnode.com/post/everything-you-need-to-know-about-socketio-ck71r30il06dyd9s1c0gbhfu6
* https://unetworkingab.medium.com/moving-from-socket-io-to-%C2%B5websockets-js-a85753b93a5a

(the ably blog is FANTASTIC for any topic related to all of this) 

However, the dev here seems to be completely absent so I may rethink using this before actually implementing this (I'm working on other things and will get to sockets sometime in 3-6 months).

**Reply by qiulang** [reply]
Actually I also read articles from alby before. Alby clearly sees socket.io as a direct competitor to their own product, so you can not take their words 100% when talking about socket.io (good articles none the less).

The reason I am interested in these may be similar to yours. I have different projects using different ws solutions, I have a project with sockeet.io, another with [ws](https://github.com/websockets/ws). I tried [sockjs](https://github.com/sockjs/sockjs-client) but gave it up.

I have not tried [uWebSockets.js](https://github.com/uNetworking/uWebSockets.js). But seeing soketi using it gives me some incentive to try.

**Reply by nickchomey** [reply]
> Alby clearly sees socket.io as a direct competitor to their own product, so you can not take their words 100% when talking about socket.io (good articles none the less).

I think Ably wrote all of those articles very objectively. In fact, [they even write tutorials](https://ably.com/blog/building-a-realtime-chat-app-with-react-laravel-and-websockets.) on how to use other tools. 

Moreover, the same skepticism would apply to uWebSockets and all of the other tools that they mention. Note that [Ably says](https://ably.com/topic/websockets#web-sockets) "to the best of my knowledge, μWS is the fastest WebSocket server implementation available by a mile. It’s actually used under the hood by SocketCluster, which I’ll talk about below." 

Clearly, for whatever reason, you've got an enormous bias towards socket.io, are just looking for someone to tell you to keep using it, and no amount of evidence will convince you otherwise. I really dont care if you do or dont. Believe Ably and the uWebSockets dev, or dont. Use their tools, or dont. Benchmark them yourself, or dont. 

Good luck with your projects.
